pkg/common/apply/rbac.go

```diff
@@ -0,0 +1,147 @@
+package apply
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/client-go/kubernetes"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+)
+
+type PermissionApplier struct {
+	roleLister               rbacv1listers.RoleLister
+	roleBindingLister        rbacv1listers.RoleBindingLister
+	clusterRoleLister        rbacv1listers.ClusterRoleLister
+	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingLister
+	client                   kubernetes.Interface
+}
+
+func NewPermissionApplier(
+	client kubernetes.Interface,
+	roleLister rbacv1listers.RoleLister,
+	roleBindingLister rbacv1listers.RoleBindingLister,
+	clusterRoleLister rbacv1listers.ClusterRoleLister,
+	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingLister,
+) *PermissionApplier {
+	return &PermissionApplier{
+		client:                   client,
+		roleLister:               roleLister,
+		roleBindingLister:        roleBindingLister,
+		clusterRoleLister:        clusterRoleLister,
+		clusterRoleBindingLister: clusterRoleBindingLister,
+	}
+}
+
+func (m *PermissionApplier) Apply(
+	ctx context.Context,
+	recorder events.Recorder,
+	manifests resourceapply.AssetFunc,
+	files ...string) []resourceapply.ApplyResult {
+	ret := []resourceapply.ApplyResult{}
+	for _, file := range files {
+		result := resourceapply.ApplyResult{File: file}
+		objBytes, err := manifests(file)
+		if err != nil {
+			result.Error = fmt.Errorf("missing %q: %v", file, err)
+			ret = append(ret, result)
+			continue
+		}
+		requiredObj, err := resourceread.ReadGenericWithUnstructured(objBytes)
+		if err != nil {
+			result.Error = fmt.Errorf("cannot decode %q: %v", file, err)
+			ret = append(ret, result)
+			continue
+		}
+		result.Type = fmt.Sprintf("%T", requiredObj)
+		switch t := requiredObj.(type) {
+		case *rbacv1.ClusterRole:
+			result.Result, result.Changed, result.Error = Apply[*rbacv1.ClusterRole](
+				ctx, m.clusterRoleLister, m.client.RbacV1().ClusterRoles(), compareClusterRole, t, recorder)
+		case *rbacv1.ClusterRoleBinding:
+			result.Result, result.Changed, result.Error = Apply[*rbacv1.ClusterRoleBinding](
+				ctx, m.clusterRoleBindingLister, m.client.RbacV1().ClusterRoleBindings(), compareClusterRoleBinding, t, recorder)
+		case *rbacv1.Role:
+			result.Result, result.Changed, result.Error = Apply[*rbacv1.Role](
+				ctx, m.roleLister.Roles(t.Namespace), m.client.RbacV1().Roles(t.Namespace), compareRole, t, recorder)
+		case *rbacv1.RoleBinding:
+			result.Result, result.Changed, result.Error = Apply[*rbacv1.RoleBinding](
```

Comment:
if the "roleRef" field changes, the apply will fail, right?

Reply:
maybe not for now. it is implicit